src/languages/rust.ts

```diff
@@ -0,0 +1,128 @@
+  "inner_attribute_item",
+  "let_declaration",
+  "macro_definition",
+  "macro_invocation",
+  "mod_item",
+  "static_item",
+  "struct_item",
+  "trait_item",
+  "type_item",
+  "union_item",
+  "use_declaration",
+  "expression_statement",
+];
+
+/**
+ * Scope types allowed to be parents of a statement
+ */
+const STATEMENT_PARENT_TYPES = ["source_file", "block", "declaration_list"];
+
+/**
+ * Returns the return value node for a given block if we are in a block that has
+ * a return value. If the return value expression uses the return keyword then
+ * we return the value itself otherwise we just return the expression
+ * @param node The node which we might match
+ * @returns The return value node
+ */
+function returnValueFinder(node: SyntaxNode) {
+  if (node.type !== "block") {
+    return null;
+  }
+
+  const { lastNamedChild } = node;
+
+  // The return expression will always be the last statement or expression in
+  // the block
+  if (lastNamedChild == null) {
+    return null;
+  }
+
+  // If the final name child is an expression statement not a raw expression
+  // then we only treat it as a return value if it is a return expression.
+  // Otherwise it is just a normal statement that doesn't return anything
+  if (lastNamedChild.type === "expression_statement") {
+    const expression = lastNamedChild.child(0)!;
+
+    if (expression.type === "return_expression") {
+      return expression.child(1);
+    }
+
+    return null;
+  }
+
+  // Any other type of statement is not a return statement so we should not
+  // match it
+  if (STATEMENT_TYPES.includes(lastNamedChild.type)) {
+    return null;
+  }
+
+  // NB: At this point we have now excluded all statement types of the only other
+  // possible node is an expression node
+
+  // If it is a return expression then we zoom down to the actual value of the
+  // return expression. This happens when they say `return foo` with no
+  // trailing semicolon
+  if (lastNamedChild.type === "return_expression") {
+    return lastNamedChild.child(1);
+  }
+
+  // At this point it is an expression which is not a return expression so we
+  // just return it as the return value of the block
+  return lastNamedChild;
+}
+
+const nodeMatchers: Partial<
+  Record<SimpleScopeTypeType, NodeMatcherAlternative>
+> = {
+  statement: STATEMENT_TYPES,
+  string: ["raw_string_literal", "string_literal"],
+  ifStatement: "if_expression",
+  functionCall: ["call_expression", "macro_invocation", "struct_expression"],
+  comment: ["line_comment", "block_comment"],
+  list: ["array_expression", "tuple_expression"],
+  collectionItem: argumentMatcher("array_expression", "tuple_expression"),
+  namedFunction: "function_item",
+  type: cascadingMatcher(
+    leadingMatcher(
+      ["let_declaration[type]", "parameter[type]", "field_declaration[type]"],
+      [":"]
+    ),
+    leadingMatcher(["function_item[return_type]"], ["->"]),
+    patternMatcher("struct_item", "trait_item", "impl_item", "impl_item[type]")
```

Comment:
Hmm looks like the order got switched back. But either way it just occurred to me that this won't actually work with either order. The problem is that with `impl_item[type]`, the node matcher is still just looking for an `impl_item`, and then returning `[type]` if that field exists. That means that this matcher will return the type no matter where you are in the `impl_item`

@AndreasArvidsson I'm assuming there's no way to do this using our pattern matchers?

Your best bet is prob to write a custom finder. Shouldn't be too bad. Basically just check if parent is `impl_item` and if the node is its `[type]` field

Reply:
You are correct

Reply:
How do I check if the node is inside his `type` field? By checking ranges?

Reply:
Write a finder where given `node`, check if `node.parent.type === `"impl_item"` and then use [`getChildNodesForFieldName`](https://github.com/cursorless-dev/cursorless/blob/81d94e2b3926499c0a85299c023fde8da66c28d3/src/util/treeSitterUtils.ts#L18): `getChildNodesForFieldName(node.parent, "type") === node`